allocations/tasks.py

```diff
@@ -0,0 +1,92 @@
+                alloc.save()
+                logger.info('successfully roll back portal db change for {}'.format(alloc.project_charge_code))
+            except Exception as re:
+                logger.exception(re)
+
+    logger.debug('need to expire {} allocations, and {} were actually expired'.format(len(expired_allocations), expired_alloc_count))
+    
+    # check projects with multiple active allocations
+    for proj in Allocation.objects.order_by().values_list('project_charge_code', flat=True).distinct():
+        project_active_allocations = list(Allocation.objects.filter(status='active', project_charge_code=proj))
+        if len(project_active_allocations) > 1:
+            logger.warning('project {} has more than one active allocations'.format(proj))
+            max_expiration_date = max([a.expiration_date for a in project_active_allocations])
+            is_active_alloc_set = False
+            for alloc in project_active_allocations:
+                if not is_active_alloc_set and alloc.expiration_date == max_expiration_date:
+                    is_active_alloc_set = True
+                else:
+                    alloc.status = 'inactive'
+                    alloc.save()
+
+    # activate allocations
+    approved_allocations = Allocation.objects.filter(status='approved', start_date__lte=now)
+    activated_alloc_count = 0
+    for alloc in approved_allocations:
+        # deactivate active allocation for the project and set status to active
+        project_active_allocations = Allocation.objects.filter(status='active', project_charge_code=alloc.project_charge_code)
+        proj_alloc = None
+        if len(project_active_allocations) > 0:
+            proj_alloc = project_active_allocations[0]
+        with transaction.atomic():
+            if proj_alloc:
+                proj_alloc.su_used = float(balance_service.call(alloc.project_charge_code)['used'])
+                proj_alloc.status = 'inactive'
+                proj_alloc.save()
+            alloc.status = 'active'
+            alloc.save()
+        try:
+            # recharge balance service
+            balance_service.recharge(alloc.project_charge_code, alloc.su_allocated)
+            activated_alloc_count = activated_alloc_count + 1
+        except Exception as be:
+            logger.exception(be)
+            # roll back portal db change
+            try: 
+                if proj_alloc:
+                    proj_alloc.su_used = None
+                    proj_alloc.status = 'active'
+                    proj_alloc.save()
+                alloc.status = 'approved'
+                alloc.save()
+                logger.info('successfully roll back portal db change for {}'.format(alloc.project_charge_code))
+            except Exception as re:
+                logger.exception(re)
+                
+    logger.debug('need to activated {} allocations, and {} were actually activated'.format(len(approved_allocations), activated_alloc_count))
```

Comment:
This is a long task, it may be easier to read if the 3 different "subtasks" are expressed as their own private function.